### DISCUSSION TITLE
Slow performance with large buckets on R2

### DISCUSSION THREAD

**Discussion by markabur** [iterate-ch/cyberduck#15719] [Q&A]
I have some buckets on R2 with several thousand files each. When I double-click the bucket in Cyberduck, it takes a long time to list the bucket contents. For example, a bucket with 3600 files took 7 minutes. A bucket with 7600 files took 31 minutes. 

**To Reproduce**
1. Have an R2 account with many thousand files per bucket
2. Double-click bucket to view contents
3. Wait

**Expected behavior**
I expect/hope performance on R2 would be about the same as Rackspace Cloud Files

**Desktop (please complete the following information):**
 - OS: Mac OS
 - Version 14.2.1

**Additional context**
Listing and uploading to a bucket with just a few files is as fast as I'd expect.

The R2 buckets are a clone of buckets we have on Rackspace Cloud Files and Cyberduck does just fine there. 

I am using an R2 profile with S3 timestamps enabled.

**Log Files**
If needed, I can supply log files or snippets. I can say that when I list my "doc13" bucket (3600 files) on Rackspace with debugging enabled, the log file is 905 KB. When I list my "doc13" bucket on R2, the log file is 31.5 MB. The big difference is that R2 has a HEAD request for every file but Rackspace doesn't.

**Comment by dkocher** [comment]
Using the S3 (Timestamps) connection profile [^1] will read metadata for *every* file to be displayed. That causes an additional request to the S3 API that will degrade performance as described. However, you only need to use the profile when you want to view the timestamps set in the browser, but they are always set with the generic S3 connection profile for files uploaded.  I have opened [^2] to fix our documentation in this regard. 

[^1]: https://docs.cyberduck.io/protocols/s3/#modification-date
[^2]: iterate-ch/docs#485

**Reply by markabur** [reply]
Ok, thanks. Would Backblaze B2 act the same way?

**Reply by dkocher** [reply]
B2 does not require an additional API request to obtain file info but includes this metadata when listing objects and is thus not affected by the same performance drawback.

**Reply by markabur** [reply]
Excellent, thank you!
